### DISCUSSION TITLE
Can it use native OpenGL 2.1 for raspberry pi instead of GL4ES?

### DISCUSSION THREAD

**Discussion by sosssego** [ptitSeb/box86#405] [Q&A]
Is it possible to use Raspberry Pi version of OpenGL 2.1 instead of GL4ES?

Also I would like to understand how it works, any suggestion on what to read?
How it translate for example DirectX calls to OpenGL?

Thanks
ps. Great work!

**Comment by ptitSeb** [comment]
> Is it possible to use Raspberry Pi version of OpenGL 2.1 instead of GL4ES?
> 
Yes of course. Box86 will use whatever libGL.so.1 it find first. On a Pi4, just don't put GL4ES to use default  GL (so mesa)

> Also I would like to understand how it works, any suggestion on what to read?

Here is an old article about box86: https://www.giantpockets.com/box86-run-x86-code-and-games-on-arm/ It's a bit old, so the Dynarec didn't exist at that time. But the principle of box86 are explained. Just not the dynarec.
Also I write a blog, so you can have some more technical info from time to time; https://box86.org/index.php/2021/06/27/game-performances/

> How it translate for example DirectX calls to OpenGL?
> 
That's Wine working there, not box86 directly. Box86 allow x86 version of Wine to run (translating x86 binaries to ARM). And Wine is doing all the Windows to Linux translation.

> Thanks
> ps. Great work!

Thanks :)

**Reply by sosssego** [reply]
Thanks for your answer, I read the 2 links they are good.
I still did not understand all of it, I think I need to study the .elf part
I did not get how you check that the one program is calling a function, but I think I should try replicate the system call example in your link. It may be I good idea put links to the commits in the article.

Regards

**Reply by ptitSeb** [reply]
Function call are indeed resolved durring the elf loading. You can use `readelf -aW /PATH/TO/SOME/ELF | less` to have some detailled informations on what an elf file needs...
